pandas/core/generic.py

```diff
@@ -1085,10 +1094,19 @@ def rename(self, *args, **kwargs):
             if v is None:
                 continue
             f = com._get_rename_function(v)
-
             baxis = self._get_block_manager_axis(axis)
             if level is not None:
                 level = self.axes[axis]._get_level_number(level)
+
+            # GH 13473
+            if not callable(v):
+                indexer = self.axes[axis].get_indexer_for(v)
+                missing_labels = [label for index, label in enumerate(v)
```

Comment:
I'm a tad concerned about perf here. Can you instead check first if there are any missing labels with something like

```python
missing = indexer[indexer == -1]
if len(missing):
    ...
```

I'd like to avoid the iteration over `v`, which may be large.

but maybe measure the relative perf of the two approaches.

Reply:
I understand what you mean; I introduced `missing_labels` to generate a nice error message, where only the labels which are missing are in the error message.

I see two options:
- Remove `missing_labels` altogether and just place `v` inside the error message.
- Move `missing_labels` into the if-statement, so that the enumeration is only done in the case of an KeyError.

I'd personally prefer the latter, what do you think is best?

Reply:
Sorry if I wasn't clear. Including the missing labels in the error message is nice. Moving it under the `if` statement is best.